packages/SwingSet/src/devices/loopbox-src.js

```diff
@@ -2,16 +2,70 @@ import { assert, details as X } from '@agoric/assert';
 import { Far } from '@agoric/marshal';
 
 export function buildRootDeviceNode(tools) {
-  const { SO, endowments } = tools;
+  const {
+    SO,
+    endowments,
+    deviceParameters,
+    getDeviceState,
+    setDeviceState,
+  } = tools;
   const { registerPassOneMessage, deliverMode } = endowments;
-  const inboundHandlers = harden(new Map());
 
-  const queuedMessages = [];
+  function makeSender(sender) {
+    return Far('sender', {
+      add(peer, msgnum, body) {
+        const oldDS = getDeviceState();
+        assert(oldDS.inboundHandlers[peer], X`unregistered peer '${peer}'`);
+        const h = oldDS.inboundHandlers[peer];
+        const newDS = { ...oldDS };
+        if (deliverMode === 'immediate') {
+          SO(h).deliverInboundMessages(
+            sender,
+            harden([[oldDS.counts[sender], body]]),
+          );
+        } else {
+          newDS.queuedMessages = Array.from(newDS.queuedMessages);
+          newDS.queuedMessages.push([h, sender, oldDS.counts[sender], body]);
+        }
+        newDS.counts = { ...newDS.counts };
+        newDS.counts[sender] += 1;
+        setDeviceState(harden(newDS));
+      },
+      remove(_peer, _msgnum) {},
+      ackInbound(_peer, _msgnum) {},
+    });
+  }
+
+  let deviceState = getDeviceState();
+  if (!deviceState) {
+    deviceState = {
+      queuedMessages: [],
+      counts: {},
+      inboundHandlers: {},
+    };
+    for (const senderName of deviceParameters.senders) {
+      deviceState.counts[senderName] = 1;
+    }
+    harden(deviceState);
+  }
+  const senderMap = new Map();
+  const senders = [];
+  for (const senderName of deviceParameters.senders) {
+    const sender = makeSender(senderName);
+    senders.push(sender);
+    senderMap.set(senderName, sender);
+  }
+  setDeviceState(harden(senders));
```

Comment:
We should explain this righteous hack in a comment:

```

// Devices don't use transcripts or orthogonal persistence, so e.g.
// `makeSender()` won't be called in every process that makes up the durable
// lifetime of the swingset (in fact it will probably only be called once).
// We abuse setDeviceState() and the `m.serialize()` it does to assign
// consistent vrefs to our Sender remotables, every time the device is
// instantiated. But we must immediately overwrite the state with data that
// omits all the Remotables, since they cannot be synthesized by
// `m.unserialize()` during the next launch.
```

Reply:
Done.
Righteous isn't quite the word I would choose, but yes, it should be documented.